vault/logical_system_activity.go

```diff
@@ -142,7 +142,12 @@ func (b *SystemBackend) handleMonthlyActivityCount(ctx context.Context, req *log
 		return logical.ErrorResponse("no activity log present"), nil
 	}
 
-	results := a.partialMonthClientCount(ctx)
+	results, err := a.partialMonthClientCount(ctx)
```

Comment:
i made this comment earlier, but it doesn't look like `results` has any info in it when `err != nil` - it might be more clear to show that results has nothing in it and put an empty `Data` object in the response in this error case

Reply:
did you have any thoughts on this?

Reply:
The Go convention for funcs that return (x,error) is to populate only one of the two return values - I'd prefer if it were exactly one, but sometimes we do return (nil,nil), which tends to cause bugs because people are so used to the more common practice of always returning exactly one non-nil value.

This carries over to our handlers as well.  If they return a non-nil error, they should be returning a nil response.  The complication is that we can return errors in two ways.  We can either return (nil, err) in which case the user will not an error message, they'll just get a 500 internal server error.  Or we can return (logical.ErrorResponse("message"), nil) in which case we'll put "message" into the response's `error` field and return a 400 bad request.  Or we can return (logical.ErrorReponse("message"), err), in which case they'll see "message" in the response's `error` field and the status code of the response will depend on err, see logical.RespondErrorCommon for specifics.

There's no general guidance about when to use these different approaches, though we should probably write some.  One tip I can offer is that if the error is an "expected" error, e.g. the user provided bad input in their request, or asked to do something we don't want to allow, it's good to provide an explanation of the problem in the response's error message.  If it's an internal failure that shouldn't have happened, that isn't predictable based on the inputs, then we usually tend more towards the (nil, err) return where they just a 500.  This is because we don't know what appear in the error, and we might wind up leaking information that the user shouldn't see.  Often in that case we'll log the error.

Anyway, in this case, when err != nil, I would simply return (nil, err).